source/common/upstream/upstream_impl.h

```diff
@@ -179,13 +180,13 @@ class HostDescriptionImplBase : virtual public HostDescription,
   // possibly benefit other related and expensive computations too (e.g.: updating subsets).
   MetadataConstSharedPtr metadata() const override {
     absl::ReaderMutexLock lock(&metadata_mutex_);
-    return metadata_;
+    return endpoint_metadata_;
   }
   void metadata(MetadataConstSharedPtr new_metadata) override {
     auto& new_socket_factory = resolveTransportSocketFactory(address(), new_metadata.get());
     {
       absl::WriterMutexLock lock(&metadata_mutex_);
-      metadata_ = new_metadata;
+      endpoint_metadata_ = new_metadata;
       // Update data members dependent on metadata.
       socket_factory_ = new_socket_factory;
     }
```

Comment:
/lgtm api

But what will happen if the locality is updated but the endpoint self isn't updated? 

/wait-any

Reply:
I'm not sure how that would be possible, but either way I don't think it would matter. Both the locality and endpoint metadata are checked every time, so the change should be reflected. LMKWYT

Reply:
Sorry, the in-place host update is not allowed is locality updated. So, you are right.